### ISSUE TITLE
[Request] Echo times in 3dinfo

### ISSUE THREAD

**Issue by jbteves** [afni/afni#132]
I tried to find this in the help text and couldn't, and `grep`'d the verbose output, so have concluded it's not there. Could we get the echo times out of `3dinfo`? Happy to help with this if needed/desired. I took a quick look through the source of `3dinfo.c` and there are no references to it, but I assume the information exists somewhere in the .HEAD file. If not, maybe we could update the spec?

**Comment by leej3** [comment]
Thanks @jbteves . Do you know of a good ME dataset online we can use to debug/add this?

**Comment by jbteves** [comment]
Hm, we have a few in .nii.gz format for tedana’s integration tests, not in AFNI format. It occurs to me that may be why... is there a convenient way to add a field? I also have some dicoms from some studies here at MUSC that I could probably get permission to make public. I can fiddle with dimon tomorrow to try and convert them. Just let me know what you’d prefer.

**Comment by leej3** [comment]
No worries about the dataset. 

Is this converted with dcm2niix? It’s likely contained in the json sidecar if it’s anywhere. Support across vendors may vary though. 

John

**Comment by yarikoptic** [comment]
is there `bids2afni` (or `bids2brik`) which would digest .nii.gz's + sidecar files to create a "proper" multi-echo BRICK/HEAD?

FWIW, you could checkout those datasets:
```shell
(git-annex)lena:~/datalad/openneuro[master]
$> datalad search multi.*echo
search(ok): /home/yoh/datalad/openneuro/ds000117 (dataset)
search(ok): /home/yoh/datalad/openneuro/ds000254 (dataset)
```

**Comment by jbteves** [comment]
@leej3 yes there's a sidecar, and it's in there. It's not that I can't find the information, it's more that I thought it would be useful for it to show up if I run `3dinfo`. 
I second @yarikoptic's question. If not, I'm definitely happy to help with that. It would be great to digest the slice timing and echo time, at a minimum.

**Comment by leej3** [comment]
Thanks @yarikoptic . I'll look through those with @afni-rickr now and see what we can do to make it easier.

**Comment by leej3** [comment]
Ok. We had a look.

A few points:

+ @yarikoptic, the datalad repo doesn't work. I had to download it with `aws s3 sync --no-sign-request s3://openneuro.org/ds000254 ds000254-download`. Is the openneuro support the most appropriate place to raise this as an issue.
+ The 254 dataset has the json in the root of the tree and and so parsing the full tree is the only way to usefully implement this sort of bids metadata consumption.
+ We don't want to depend on pybids, nor do we want to reinvent the wheel internally.

The bids2afni seems like an appealing option. We would happily maintain a dictionary of bids json metadata fields to afni's attributes in the niml object. I'm also happy to help write such a tool. We could use pybids to parse the bids tree, and use the above mentioned dictionary to populate the output tree with data/metadata in which all metadata associated with a file is in its unique sidecar. 

A key question would be whether we could separate the two tasks tasks 
a)  maintaining a map of metadata names (specific to AFNI and could be maintained largely by us in the afni-codebase, or a separate repo)
b) writing out such a tree has the implicit inheritance structure of the bids tree metadata removed, something that may be useful to all software that does not wish to support the inheritance structure of the metadata.

**Comment by jbteves** [comment]
Personally, I'd be pretty happy with the `bids2afni` either requiring a `.json` of the same name as the input or requiring that the user pass one, and it just being a dictionary in the `.HEAD`. Then if we want to get fancy, `tedana` could check the echo times inside of that dictionary. 
I don't think AFNI necessarily needs to support creating a whole tree unless it's trying to be strictly BIDS-adherent. Then again, I am biased because I'm a "no inheritance" proponent following discussions with some of the Code Convergence attendees and personal experience trying to understand the different cases within the standard. If AFNI did create such a tool, though I would vote to sarcastically call it `3ddeathtax` since it takes away inheritance, effectively.

**Comment by afni-rickr** [comment]
There are 2 pieces to this that we are picturing.  The `bids2afni` piece of trickling down metadata from the branch root JSON files down to the lowly leaves (JSON files) is not quite AFNI specific.  Things like echo and slice timing should be mostly desirable across packages.  I am not sure what AFNI might want that other packages would not, say.
The other piece is handling of the JSON side cars (the parallel ones), which would be done internally (though probably with at least on/off control via an env var).  We would probably not make a special tool for that, it would just happen.

**Comment by jbteves** [comment]
I had figured they were AFNI-specific in that they would be put into HEAD files, but if you trickle the information down, and it's not AFNI-specific, do you mean appending to the leaf JSON files from their parents?
Please forgive me if I'm misunderstanding the differences between the JSON, HEAD, NIML, and environment variables, I'm less familiar with the details there.

**Comment by jbteves** [comment]
I'll also note that ultimately, if AFNI give me a way to get the echo times via 3dinfo for convenience, the implementation details are irrelevant to close this issue. This conversation seems to have grown into something larger than that by accident. @leej3 should we open a separate one to discuss JSON inheritance, etc., in AFNI?

**Comment by afni-rickr** [comment]
Getting the TE is indeed a larger issue, because 1) it is not yet part of the AFNI header, and 2) we have to write a reasonable infrastructure for reading a JSON sidecar.  We should do both things, but I am not sure how soon it might happen.  Well, if you really, really want it, that does help us prioritize.  But yes, the JSON inheritance discussion should be separated.

**Comment by leej3** [comment]
> I had figured they were AFNI-specific in that they would be put into HEAD files, but if you trickle the information down, and it's not AFNI-specific, do you mean appending to the leaf JSON files from their parents?

Yes. The tool/pybids method would be "trickle down", "collapse inheritance" or something along those lines. After application, a tool would not have to search the full tree to fully instantiate the metadata associated with a single scan

> if I'm misunderstanding the differences between the JSON, HEAD, NIML, and environment variables, I'm less familiar with the details there.

The json can have arbitrary information in it. The .HEAD file contains the internal metadata for a dataset that AFNI uses. Some of this metadata maps directly to values stored in the nifti header. What does not map to a nifti value, is written to the AFNI extension within the nifti header. This metadata dump is an xml like format used internally by AFNI called niml.

> I'll also note that ultimately, if AFNI give me a way to get the echo times via 3dinfo for convenience

I'd prefer to solve this in a more general way so that there is a clear path to incorporating additional bids values into the list of scan metadata that we make use of. I'll open an issue on pybids to see how they feel about this. Depending on how that goes we can close this issue. Thanks for raising it. If it takes to long to solve in a generic way we can use the quick hack mentioned above to solve this specific issue.

**Comment by jbteves** [comment]
Okay, I think I follow now. 
I see. I'm fine with either of the following:
1. You guys get to it when you get to it. You're busy people with important software engineering to do, and for now I can just cat and grep the sidecar. 
1. If you help me understand how I can contribute, I can do my best to help write it into AFNI. I have some rudimentary C knowledge and okay-ish Python knowledge. As we head into the holidays, there's usually a work lull that I can leverage to get some pet projects done and this could be one of them. I can see where this would cost you more time than it saves you, however. 
I agree with you both that solving the problem generally is a much better solution. I hadn't considered the full scope of importing other metadata into AFNI. No need to thank me for raising an issue: I am, to loosely quote @yarikoptic, an "expert whiner." Feel free to tag me on the pybids issue.

**Comment by neurolabusc** [comment]
@jbteves sorry I missed this. If your DICOMs were converted to NIfTI using dcm2niix or dicm2nii, the echo time should be in the NIfTI [descrip](https://nifti.nimh.nih.gov/pub/dist/src/niftilib/nifti1.h) field, e.g. `TE=60;Time=153509.335;phase=1`, as well as in the BIDS sidecar (the `EchoTime` field). Since the BIDS standard declares [multi-echo data MUST be split into one file per echo](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/01-magnetic-resonance-imaging-data.html), you should have one file per echo, with each files `descrip` providing the TE. Just be aware many tools (FSL, SPM) reset the descrip field when they process a NIfTI image, so it might be worth inspecting the field before any other tool touches it.

**Comment by jbteves** [comment]
@neurolabusc yes, thanks! Fortunately it is in the .json file and correct whenever I've used `dcm2niix`. I was unaware of the file-stripping, I will have to keep a closer eye on that.
